### DISCUSSION TITLE
question/what-is-closure-and-how-does-it-work

### DISCUSSION THREAD

**Discussion by giscus** [us-fe/us-fe.github.io#29] [Announcements]
# question/what-is-closure-and-how-does-it-work

北美前端面试攻略

https://us-fe.github.io/question/what-is-closure-and-how-does-it-work.html

**Comment by xiaoyu2er** [comment]
https://whatthefuck.is/closure by dan
> You have a closure when a function accesses variables defined outside of it.

**Reply by xiaoyu2er** [reply]
https://us-fe.github.io/ebooks/en/you-dont-know-js-v1/scope-closures/ch5.html#review-tl-dr

> Closure is when a function can remember and access its lexical scope even when it's invoked outside its lexical scope.

**Reply by xiaoyu2er** [reply]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures

> A closure is the combination of a function bundled together (enclosed) with references to its surrounding state (the lexical environment). In other words, a closure gives you access to an outer function's scope from an inner function. In JavaScript, closures are created every time a function is created, at function creation time.

**Comment by xiaoyu2er** [comment]
https://stackoverflow.com/questions/111102/how-do-javascript-closures-work

> A closure is a pairing of:
> + A function and
> + A reference to that function's outer scope (lexical environment)

> A lexical environment is part of every execution context (stack frame) and is a map between identifiers (i.e. local variable names) and values.

> Every function in JavaScript maintains a reference to its outer lexical environment. This reference is used to configure the execution context created when a function is invoked. This reference enables code inside the function to "see" variables declared outside the function, regardless of when and where the function is called.
